kmk/modules/dynamic_sequences.py

```diff
@@ -0,0 +1,253 @@
+        self.status = SequenceStatus.PLAYING
+        self.start_time = ticks_ms()
+        self.index = 0
+        self.current_repetition = 0
+
+    def _stop_sequence(self, key, keyboard, *args, **kwargs):
+        if self.status == SequenceStatus.RECORDING:
+            self.stop_recording()
+        elif self.status == SequenceStatus.SET_INTERVAL:
+            self.stop_config()
+        self.status = SequenceStatus.STOPPED
+
+        # Change sequences here because stop is always called
+        if key.meta.sequence_select is not None:
+            self.current_slot = self.sequences[key.meta.sequence_select]
+
+    # Configure repeat settings
+    def _set_sequence_repetitions(self, key, keyboard, *args, **kwargs):
+        self._stop_sequence(key, keyboard)
+        self.status = SequenceStatus.SET_REPEPITIONS
+        self.last_config_frame = set()
+        self.current_slot.repetitions = 0
+        self.start_time = ticks_ms()
+
+    def _set_sequence_interval(self, key, keyboard, *args, **kwargs):
+        self._stop_sequence(key, keyboard)
+        self.status = SequenceStatus.SET_INTERVAL
+        self.last_config_frame = set()
+        self.current_slot.interval = 0
+        self.start_time = ticks_ms()
+
+    # Add the current keypress state to the sequence
+    def record_frame(self, keys_pressed):
+        if self.current_slot.sequence_data[self.index][0] != keys_pressed:
+            self.index += 1
+
+            # Recorded speed
+            if self.use_recorded_speed:
+                self.current_slot.sequence_data.append(
+                    (keys_pressed.copy(), ticks_diff(ticks_ms(), self.start_time))
+                )
+
+            # Constant speed
+            else:
+                self.current_slot.sequence_data.append(
+                    (keys_pressed.copy(), self.index * self.key_interval)
+                )
+
+        if not check_deadline(ticks_ms(), self.start_time, self.timeout):
+            self.stop_recording()
+
+    # Add the ending frames to the sequence
+    def stop_recording(self):
+        # Clear the remaining keys
+        self.current_slot.sequence_data.append(
+            (set(), self.current_slot.sequence_data[-1][1] + 20)
+        )
+
+        # Wait for the specified interval
+        self.current_slot.sequence_data.append(
+            (
+                set(),
+                self.current_slot.sequence_data[-1][1]
+                + self.current_slot.interval * 1000,
+            )
+        )
+
+        self.status = SequenceStatus.STOPPED
+
+    def play_frame(self, keyboard):
+        # Send the keypresses at this point in the sequence
+        if not check_deadline(
+            ticks_ms(), self.start_time, self.current_slot.sequence_data[self.index][1]
+        ):
+            if self.index:
+                for key in self.current_slot.sequence_data[self.index - 1][
+                    0
+                ].difference(self.current_slot.sequence_data[self.index][0]):
```

Comment:
There are a couple of hard to read multi line statements. If it's that complicated, give it a name:
```suggestion
                prev = self.current_slot.sequence_data[self.index - 1][0]
                cur = self.current_slot.sequence_data[self.index][0]
                for key in prev.difference(cur):
```